dipy/denoise/localpca.py

```diff
@@ -0,0 +1,114 @@
+            for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+                for i in range(patch_radius, arr.shape[0] - patch_radius, 1):
+
+                    X = np.zeros(
+                        (patch_size * patch_size * patch_size, arr.shape[3]))
+                    M = np.zeros(arr.shape[3])
+
+                    temp = arr[i - patch_radius: i + patch_radius + 1,
+                               j - patch_radius: j + patch_radius + 1,
+                               k - patch_radius: k + patch_radius + 1, :]
+                    X = temp.reshape(
+                        patch_size *
+                        patch_size *
+                        patch_size,
+                        arr.shape[3])
+                    # compute the mean and normalize
+                    M = np.mean(X, axis=0)
+                    X = X - np.array([M, ] * X.shape[0],
+                                     dtype=np.float64)
+
+                    # Compute the covariance matrix C = X_transpose X
+                    C = np.transpose(X).dot(X)
+                    C = C / X.shape[0]
+
+                    [d, W] = np.linalg.eigh(C)
+
+                    d[d < tou[i, j, k, :]] = 0
+
+                    W_hat = W * 0
+                    W_hat[:, d > 0] = W[:, d > 0]
+                    Y = X.dot(W_hat)
+                    X_est = Y.dot(np.transpose(W_hat))
+
+                    # theta value
+                    temp = X_est + \
+                        np.array([M, ] * X_est.shape[0], dtype=np.float64)
+                    temp = temp.reshape(
+                        patch_size, patch_size, patch_size, arr.shape[3])
+                    # Also update the estimate matrix which is X_est * theta
+
+                    theta[i - patch_radius: i + patch_radius + 1,
+                          j - patch_radius: j + patch_radius + 1,
+                          k - patch_radius: k + patch_radius + 1,
+                          :] = theta[i - patch_radius: i + patch_radius + 1,
+                                     j - patch_radius: j + patch_radius + 1,
+                                     k - patch_radius: k + patch_radius + 1,
+                                     :] + 1.0 / (1.0 + np.linalg.norm(d,
+                                                                      ord=0))
+
+                    thetax[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1,
+                           :] = thetax[i - patch_radius: i + patch_radius + 1,
+                                       j - patch_radius: j + patch_radius + 1,
+                                       k - patch_radius: k + patch_radius + 1,
+                                       :] + temp / (1 + np.linalg.norm(d,
+                                                                       ord=0))
+
+        denoised_arr = thetax / theta
```

Comment:
What I actually found is that the adaptation step (eq 5 and 6) worsens the results and it doesn't mat at all with the Manjon's MATLAB implementation. We compared our results withe their output with the current algorithm and it is really really close. Plus even though the MATLAB implementation has an option to choose between the Gaussian and Rician Noise, the error map between the two output is exactly zero, which makes us believe that they might have used is minus the debias step.